File: src/auth/onboarding.py
"""Onboarding state machine for the self-serve hosted flow."""

# ---------------------------------------------------------------------------
# Onboarding progress states  (users.onboarding_status)
# ---------------------------------------------------------------------------

PENDING_EMAIL_VERIFICATION = "pending_email_verification"
PENDING_DB_CONNECTION = "pending_db_connection"
SETUP_COMPLETE = "setup_complete"

# ---------------------------------------------------------------------------
# Account states  (users.account_status)
# ---------------------------------------------------------------------------

ACCOUNT_ACTIVE = "active"
ACCOUNT_SUSPENDED = "suspended"
ACCOUNT_CLOSED = "closed"

ACCOUNT_STATES: frozenset[str] = frozenset({ACCOUNT_ACTIVE, ACCOUNT_SUSPENDED, ACCOUNT_CLOSED})
TERMINAL_ACCOUNT_STATES: frozenset[str] = frozenset({ACCOUNT_CLOSED})

# ---------------------------------------------------------------------------
# Billing states  (users.billing_status)
# ---------------------------------------------------------------------------

BILLING_FREE = "free"
BILLING_TRIALING = "trialing"
BILLING_ACTIVE_PAID = "active_paid"
BILLING_PAST_DUE = "past_due"
BILLING_CANCELED = "canceled"

# ---------------------------------------------------------------------------
# Backward-compat aliases  prefer ACCOUNT_* in new code
# ---------------------------------------------------------------------------

ACTIVE = ACCOUNT_ACTIVE
SUSPENDED = ACCOUNT_SUSPENDED
CLOSED = ACCOUNT_CLOSED

# ---------------------------------------------------------------------------
# Triggers
# ---------------------------------------------------------------------------

TRIGGER_EMAIL_VERIFIED = "email_verified"
TRIGGER_DB_SUBMITTED = "db_submitted"

# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------


class InvalidTransitionError(Exception):
    """Raised when (current_state, trigger) has no valid destination."""


# ---------------------------------------------------------------------------
# Next-step descriptions
# ---------------------------------------------------------------------------

_NEXT_STEP_DESCRIPTIONS: dict[str, str] = {
    PENDING_EMAIL_VERIFICATION: "Check your email and click the verification link.",
    PENDING_DB_CONNECTION: "Submit your database connection details.",
    SETUP_COMPLETE: "Setup complete. Create an API key to start querying.",
    ACCOUNT_ACTIVE: "Your account is active. You may use your API key.",
    ACCOUNT_SUSPENDED: "Account suspended. Contact support.",
    ACCOUNT_CLOSED: "Account closed.",
}


def get_next_step_description(status: str) -> str:
    """Return a human-readable instruction for the given onboarding or account status."""
    return _NEXT_STEP_DESCRIPTIONS.get(status, "Contact support for assistance.")


# ---------------------------------------------------------------------------
# State machine  (onboarding progress only)
# ---------------------------------------------------------------------------


def resolve_next_state(current_state: str, trigger: str) -> str:
    """Return the destination onboarding state for (current_state, trigger).

    Only handles onboarding progress transitions. Account-level state changes
    (suspend, close) are managed directly through UserStore.

    Raises InvalidTransitionError when the transition is not allowed.
    """
    if current_state == PENDING_EMAIL_VERIFICATION and trigger == TRIGGER_EMAIL_VERIFIED:
        return PENDING_DB_CONNECTION

    if current_state == PENDING_DB_CONNECTION and trigger == TRIGGER_DB_SUBMITTED:
        return SETUP_COMPLETE

    raise InvalidTransitionError(f"No transition from '{current_state}' on trigger '{trigger}'.")

File: frontend/src/lib/onboarding.ts
import type {
  AccountStatus,
  AccountStatusResponse,
  OnboardingBlocker,
  OnboardingStatus,
} from '@/types/api'

export type ProtectedRoute = '/api-keys' | '/setup/api-key' | '/setup/clients' | '/setup/database' | '/setup/status'

export function resolveOnboardingDestination(
  status: OnboardingStatus,
  accountStatus: AccountStatus = 'active',
): ProtectedRoute {
  if (accountStatus !== 'active') {
    return '/setup/status'
  }

  switch (status) {
    case 'setup_complete':
      return '/setup/clients'
    case 'pending_db_connection':
      return '/setup/database'
    case 'pending_email_verification':
    default:
      return '/setup/status'
  }
}

export function resolveStatusResponseDestination(
  payload: Pick<AccountStatusResponse, 'status' | 'account_status'>,
): ProtectedRoute {
  return resolveOnboardingDestination(payload.status, payload.account_status)
}

export function getStatusPageCopy(
  status: OnboardingStatus,
  accountStatus: AccountStatus,
  blockers: OnboardingBlocker[],
): { title: string; detail: string } {
  if (accountStatus === 'suspended' || blockers.includes('account_suspended')) {
    return {
      title: 'Account suspended',
      detail: 'Your account is suspended. You will not be able to continue setup until the restriction is cleared.',
    }
  }

  if (accountStatus === 'closed' || blockers.includes('account_closed')) {
    return {
      title: 'Account closed',
      detail: 'This account has been closed. Setup and API-key management are no longer available.',
    }
  }

  if (status === 'pending_email_verification' || blockers.includes('email_verification')) {
    return {
      title: 'Email verification required',
      detail: 'Your email address must be verified before setup can continue.',
    }
  }

  return {
    title: 'Setup unavailable',
    detail: 'This account cannot continue setup from the current state.',
  }
}